tensorflow/compiler/mlir/tosa/transforms/legalize_tfl.cc

```diff
@@ -177,6 +178,106 @@ DECL_CONVERT_OP(FakeQuant);
+        op, "input/output are not all a ranked tensor");
+  }
+
+  UniformQuantizedType in_quant_type =
+      input_type.getElementType().dyn_cast<mlir::quant::UniformQuantizedType>();
+  UniformQuantizedType out_quant_type =
+      output_type.getElementType()
+          .dyn_cast<mlir::quant::UniformQuantizedType>();
+
+  if ((in_quant_type == nullptr) != (out_quant_type == nullptr)) {
+    return rewriter.notifyMatchFailure(
+        op,
+        "input/output tensor should be all quantized or all floating-point");
+  }
+
+  if (out_quant_type) {
+    // The formal definition of gelu.
+    auto gelu_func = [](double x) -> double {
+      return 0.5 * x * (1.0 + std::erf(x / std::sqrt(2)));
+    };
+
+    if (in_quant_type.getStorageTypeIntegralWidth() != 8) {
+      return rewriter.notifyMatchFailure(
+          op, "current tfl.gelu only support 8-bit quantized type");
+    }
+
+    Value table_const = getTosaConst8bitTable(
+        rewriter, op, in_quant_type.getScale(), in_quant_type.getZeroPoint(),
+        out_quant_type.getScale(), out_quant_type.getZeroPoint(), gelu_func);
+
+    CreateReplaceOpAndInfer<tosa::TableOp>(rewriter, op, output_type, input,
+                                           table_const);
+    return success();
+  }
+
+  // Following approximated implemention described in
+  // tensorflow/lite/kernels/internal/reference/gelu.h
+  // gelu(x) = 0.5 * x * (1 + tanh(sqrt(2/pi) * (x + 0.044715 * x^3)))
+  // Lower the formula to the sequence of operators below.
+  // op0 = pow(x, 3)
+  // op1 = mul(op0, 0.044715)
+  // op2 = add(x, op1)
+  // op3 = mul(op2, sqrt(2/pi))
+  // op4 = tanh(op3)
+  // op5 = add(op4 ,1)
+  // op6 = mul(x, 0.5)
+  // op7 = mul(op6, op5)
+
+  auto fp_scalar_ty = RankedTensorType::get({}, rewriter.getF32Type());
+
+  Value cst_3 = rewriter.create<tosa::ConstOp>(
+      loc, fp_scalar_ty, DenseElementsAttr::get(fp_scalar_ty, {3.000000e+00f}));
+  auto op0_pow_3 =
+      CreateOpAndInfer<tosa::PowOp>(rewriter, loc, output_type, input, cst_3);
+
+  Value cst_004 = rewriter.create<tosa::ConstOp>(
+      loc, fp_scalar_ty, DenseElementsAttr::get(fp_scalar_ty, {4.471500e-02f}));
+  auto op1_mul_op0_004 = CreateOpAndInfer<tosa::MulOp>(
+      rewriter, loc, output_type, op0_pow_3, cst_004, 0);
+
+  auto op2_add_x_op1 = CreateOpAndInfer<tosa::AddOp>(rewriter, loc, output_type,
+                                                     input, op1_mul_op0_004);
+
+  Value cst_sqrt2pi = rewriter.create<tosa::ConstOp>(
+      loc, fp_scalar_ty, DenseElementsAttr::get(fp_scalar_ty, {0.797884583f}));
+  auto op3_mul_op2_sqrt2pi = CreateOpAndInfer<tosa::MulOp>(
+      rewriter, loc, output_type, op2_add_x_op1, cst_sqrt2pi, 0);
+
+  auto op4_tanh_op3 = CreateOpAndInfer<tosa::TanhOp>(rewriter, loc, output_type,
+                                                     op3_mul_op2_sqrt2pi);
+
+  Value cst_1 = rewriter.create<tosa::ConstOp>(
+      loc, fp_scalar_ty, DenseElementsAttr::get(fp_scalar_ty, {1.000000e+00f}));
```

Comment:
Nit: drop 0's here. 1.0f, 3.0f and 0.5f read a bit more natural.

Reply:
Yes, it is. Done.